main/src/ca/uwaterloo/flix/language/phase/optimizer/LambdaDrop.scala

```diff
@@ -75,14 +75,15 @@ object LambdaDrop {
     * A function is droppable if
     * (a) it contains at least one recursive call and
     * (b) is a higher-order function, i.e., it has at least one formal parameter
-    * with a function type.
+    * with a function type and
+    * (c) has at least one constant parameter.
     *
     * The latter condition can be checked by the [[isHigherOrder]] predicate.
     */
   private def isDroppable(defn: MonoAst.Def)(implicit lctx: LocalContext): Boolean = {
     if (isHigherOrder(defn)) {
       visitExp(defn.exp)(defn.sym, lctx)
-      lctx.recursiveCalls.nonEmpty
+      lctx.recursiveCalls.nonEmpty && hasConstantParameter(lctx.recursiveCalls.toList, defn.spec.fparams)
```

Comment:
`hasConstantParameter` implicitly checks that lctx.recursiveCalls is non-empty, so we could in principle remove that check. However, I think it makes the code more readable, since it corresponds to the documentation. Additionally, if it is empty, then `hasConstantParameter` is never called so we save a few instructions that just return `Nil` and `false` (not that it really matters)